### PR TITLE
[docs] Randomness content

### PR DESCRIPTION
**Pull request by ronny-mysten**
## Description 

Adds an on-chain randomness doc.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: docs/content/guides/developer/advanced/randomness-onchain.mdx [hunk 1]
```diff
@@ -0,0 +1,140 @@
+---
+title: Access On-Chain Randomness
```
Comment (benr-ml) [inline docs/content/guides/developer/advanced/randomness-onchain.mdx:2]:
```suggestion
title: On-Chain Randomness
```

File: docs/content/guides/developer/advanced/randomness-onchain.mdx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+description: Randomness is a valuable tool to simulate chance on chain, but can also expose flaws in your logic. Understanding the vulnerabilities and accounting for them can mitigate the threat exposure for your smart contracts. 
+---
+
+Generating pseudo-random values in Move is similar to solutions in other languages. A Move function can create a new instance of `RandomGenerator` and use it for generating random values of different types, for example, `generate_u128(&mut generator), generate_u8_in_range(&mut generator, 1, 6)`, or,
+
+```rust
+entry fun roll_dice(r: &Random, ctx: &mut TxContext): Dice {
+    let generator = new_generator(r, ctx); // generator is a PRG
+		Dice { value: random::generate_u8_in_range(&mut generator, 1, 6) }
```
Comment (benr-ml) [inline docs/content/guides/developer/advanced/randomness-onchain.mdx:11]:
```suggestion
    Dice { value: random::generate_u8_in_range(&mut generator, 1, 6) }
```

File: docs/content/guides/developer/advanced/randomness-onchain.mdx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+- Define your function as (private) `entry`.
+- Prefer generating randomness using function-local `RandomGenerator`.
+- Make sure that the "unhappy path" of your function does not charge more gas than the "happy path".
+
+## Use (non-public) `entry` functions
+
+While composition is very powerful for smart contracts, it opens the door to attacks on functions that use randomness. Consider for example the next module:
+
+```rust
+module games::dice {
```
Comment (benr-ml) [inline docs/content/guides/developer/advanced/randomness-onchain.mdx:28]: indentation?

File: docs/content/guides/developer/advanced/randomness-onchain.mdx [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+:::
+
+## Instantiating `RandomGenerator`
+
+`RandomGenerator` is secure as long as it's created by the consuming module. If passed as an argument, the caller might be able to predict the outputs of that `RandomGenerator` instance (for example, by calling `bcs::to_bytes(&generator)` and parsing its internal state).
+
+:::note
+
+The Move compiler prevents defining `public` functions with `RandomGenerator` as an argument.
```
Comment (benr-ml) [inline docs/content/guides/developer/advanced/randomness-onchain.mdx:98]: maybe let's be consistent with the note from above ("Move compiler enforces this behavior by rejecting public functions with Random as an argument."), either that option or the one here in both

File: docs/content/guides/developer/advanced/randomness-onchain.mdx [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+...
+}
+```
+
+An attacker can deploy the next function:
+
+```move
+public fun attack(guess: u8, r: &Random, ctx: &mut TxContext): GuessedCorrectly {
+		let output = dice::play_dice(guess, r, ctx);
+		option::extract(output) // reverts the transaction if roll_dice returns option::none()
```
Comment (benr-ml) [inline docs/content/guides/developer/advanced/randomness-onchain.mdx:54]:
```suggestion
    option::extract(output) // reverts the transaction if play_dice returns option::none()
```

File: docs/content/guides/developer/advanced/randomness-onchain.mdx [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	}
+...
+}
+```
+
+An attacker can deploy the next function:
+
+```move
+public fun attack(guess: u8, r: &Random, ctx: &mut TxContext): GuessedCorrectly {
+		let output = dice::play_dice(guess, r, ctx);
```
Comment (benr-ml) [inline docs/content/guides/developer/advanced/randomness-onchain.mdx:53]:
```suggestion
   let output = dice::play_dice(guess, r, ctx);
```

File: docs/content/guides/developer/advanced/randomness-onchain.mdx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+}
+```
+
+The attacker can now call `attack` with a guess, and always revert the fee transfer if the guess is incorrect.
+
+To protect against composition attacks in this example, define `play_dice` as a private `entry` function so functions from other modules cannot call it, such as, 
+
+```move
+entry fun play_dice(guess: u8, fee: Coin<SUI>, r: &Random, ctx: &mut TxContext): Option<GuessedCorrectly> {
+		...
```
Comment (benr-ml) [inline docs/content/guides/developer/advanced/randomness-onchain.mdx:64]:
```suggestion
    ...
```

File: docs/content/guides/developer/advanced/randomness-onchain.mdx [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+:::
+
+## Programmable transaction block (PTB) restrictions
+
+A similar attack to the one previously described involves PTBs *even* when `play_dice` is defined as a private `entry` function. For example, consider the `entry play_dice(guess: u8, fee: Coin<SUI>, r: &Random, ctx: &mut TxContext): Option<GuessedCorrectly> { … }` function defined earlier, the attacker can publish the function
+
+```move
+public fun attack(output: Option<GuessedCorrectly>): GuessedCorrectly {
+		option::extract(output)
```
Comment (benr-ml) [inline docs/content/guides/developer/advanced/randomness-onchain.mdx:80]:
```suggestion
    option::extract(output)
```

File: docs/content/guides/developer/advanced/randomness-onchain.mdx [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+The Move compiler prevents defining `public` functions with `RandomGenerator` as an argument.
+
+:::
+
+## Limited resources and `Random` dependent flows
+
+Developers should be aware that some resources that are available to transactions are limited. If a function that reads `Random` consumes more resources in the unhappy path flow than in the happy path flow, an attacker can use that difference to revert the transaction in the unhappy flow as previously demonstrated. Concretely, gas is such a resource. Consider the following code:
+
+```rust
+entry fun play(r: &Random, payment: Coin<SUI>, ...) {
```
Comment (benr-ml) [inline docs/content/guides/developer/advanced/randomness-onchain.mdx:107]: indentation needs an update

File: docs/content/guides/developer/advanced/randomness-onchain.mdx [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+Other limited resources per transaction are:
+
+- The number of new objects.
+- The number of objects that can be used (including dynamic fields).
+
+## Related Links
+
+- [Sui Framework Reference](../../../references/framework)
+	- <a href="/references/framework/sui-framework/random" data-noBrokenLinkCheck='true'>random.move</a>
+- [Randomness example](https://github.com/MystenLabs/sui/blob/main/sui_programmability/examples/nfts/sources/random_nft.move)
```
Comment (benr-ml) [inline docs/content/guides/developer/advanced/randomness-onchain.mdx:139]: another one to add here (simpler)

Comment (ronny-mysten) [reply docs/content/guides/developer/advanced/randomness-onchain.mdx:139]: Indention? Intentional because random.move is a child of framework reference

**Review by benr-ml — APPROVED** [review]

**Pull request merged**